### PR TITLE
Cars project done!

### PR DESCRIPTION
**Pull request by pwiatr3**

### CODE REVIEW

**Comment by ziobron** [comment]
You should deliver your changes to diamond branch, not master.

File: Car.cpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+#include <iostream>
+
+Car::Car() {} 
+// Car::Car(Engine* engine)
+//     : engine_(engine)
+// {
+//     std::cout << __FUNCTION__ << std::endl;
+// }
+
+//Car::~Car()       { std::cout << __FUNCTION__ << std::endl; }
```
Comment (nauka-programowania-MA) [inline Car.cpp:11]: remove commented code!

File: Car.hpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+class Car
+{
+public:
+    Car();
+    //~Car();
```
Comment (nauka-programowania-MA) [inline Car.hpp:7]: remove

File: Car.hpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    void turnLeft();
+    void turnRight();
+    void brake();
+    void accelerate(int speed);
+    virtual void refill() = 0;
+    virtual ~Car() {};
```
Comment (nauka-programowania-MA) [inline Car.hpp:13]: Rule of 5! and also you should mark it as default

File: Car.hpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    // virtual void charge() = 0;
+    // virtual void refuel() = 0;
```
Comment (nauka-programowania-MA) [inline Car.hpp:15]: remove

File: ElectricCar.cpp [hunk 5]
```diff
@@ -8,9 +8,11 @@ ElectricCar::ElectricCar(ElectricEngine* engine)
...
 ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
 void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
-
+void ElectricCar::changeEngine(ElectricEngine* engine) {
+    delete engine_;
+    engine_ = engine;
```
Comment (nauka-programowania-MA) [inline ElectricCar.cpp:14]: use unique_ptr don't manage resources on heap manually this is not 90's :D

File: ElectricCar.hpp [hunk 6]
```diff
@@ -1,16 +1,16 @@
+#pragma once
 #include "ElectricEngine.hpp"
+#include "Car.hpp"
 
-class ElectricCar
+class ElectricCar : virtual public Car
 {
 public:
     ElectricCar(ElectricEngine* engine);
     ~ElectricCar();
```
Comment (nauka-programowania-MA) [inline ElectricCar.hpp:9]: Rule of 5!

File: HybridCar.hpp [hunk 7]
```diff
@@ -1,19 +1,14 @@
...
 #include "ElectricEngine.hpp"
+#include "PetrolCar.hpp"
 #include "PetrolEngine.hpp"
 
-class HybridCar
+class HybridCar : public ElectricCar, public PetrolCar
 {
 public:
     HybridCar(PetrolEngine* petrolEng, ElectricEngine* electricEng);
     ~HybridCar();
```
Comment (nauka-programowania-MA) [inline HybridCar.hpp:11]: Rule of 5!

File: PetrolCar.cpp [hunk 8]
```diff
@@ -8,9 +8,11 @@ PetrolCar::PetrolCar(PetrolEngine* engine)
...
-void PetrolCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
 void PetrolCar::refuel()        { std::cout << __FUNCTION__ << std::endl; }
-
+void PetrolCar::changeEngine(PetrolEngine* engine) {
+    delete engine_;
+    engine_ = engine;
+}
```
Comment (nauka-programowania-MA) [inline PetrolCar.cpp:15]: unqiue_ptr

File: PetrolCar.hpp [hunk 9]
```diff
@@ -1,16 +1,17 @@
+#pragma once
 #include "PetrolEngine.hpp"
+#include "Car.hpp"
 
-class PetrolCar
+class PetrolCar : virtual public Car
 {
 public:
     PetrolCar(PetrolEngine* engine);
     ~PetrolCar();
```
Comment (nauka-programowania-MA) [inline PetrolCar.hpp:9]: Rule of 5!

File: main.cpp [hunk 10]
```diff
@@ -1,28 +1,35 @@
...
     std::cout << std::endl << "TOYOTA" << std::endl;
     HybridCar toyota(new PetrolEngine(80, 1400, 5), new ElectricEngine(100, 540));
-    toyota.accelerate(100);
-    toyota.brake();
-    toyota.charge();
-    toyota.refuel();
+    car = &toyota;
+    car->accelerate(100);
+    car->brake();
+    car->refill();
```
Comment (nauka-programowania-MA) [inline main.cpp:34]: Write some test :)

**Review by nauka-programowania-MA — APPROVED** [review]

**Pull request closed without merging**